discovery/dockerswarm/dockerswarm.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerswarm
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/go-kit/kit/log"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	swarmLabel = model.MetaLabelPrefix + "dockerswarm_"
+)
+
+// DefaultSDConfig is the default Docker Swarm SD configuration.
+var DefaultSDConfig = SDConfig{
+	RefreshInterval: model.Duration(60 * time.Second),
+	Host:            "http://127.0.0.1:2375/",
+	Kind:            "services",
+	Port:            80,
+}
+
+type Filter struct {
+	Name   string   `yaml:"name"`
+	Values []string `yaml:"values"`
+}
+
+// SDConfig is the configuration for Docker Swarm based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+
+	Host    string `yaml:"host"`
+	url     *url.URL
+	Filters []Filter `yaml:"filters"`
```

Comment:
Same question as before, are these actually needed for performance? Filters like these should only exist when users are running into real-world performance issues, not as an alternative to relabelling. Especially when there's multiple kinds that interact, as that get's complicated as you can see in k8.

Reply:
Yes. The Swarm API's are not paginated and you can go to very big swarm clusters.

Also, filter is common for all the kinds.

Reply:
If that's the case, then we probably need to approach this like k8 and have a per-type filter.

I'd still prefer we wait for users to bring this up, we only got the k8 one due to an odd non-Prometheus use case from Loki - it was never really a problem for general users.